### PR TITLE
Field and Scan Selection

### PR DESCRIPTION
**Pull request by sjperkins**
Select on Field and Scan

### CODE REVIEW

File: xova/apps/xova/averaging.py [hunk 1]
```diff
@@ -142,19 +162,35 @@ def average_main(main_ds, time_bin_secs, chan_bin_size,
...
     for ds in main_ds:
-        if respect_flag_row is False:
-            ds = ds.assign(FLAG_ROW=(("row",), ds.FLAG.data.all(axis=(1, 2))))
+        compress = False
+
+        if fields and ds.FIELD_ID not in fields:
+            compress = True
+
+        if scan_numbers and ds.SCAN_NUMBER not in scan_numbers:
+            compress = True
```
Comment (sjperkins) [inline xova/apps/xova/averaging.py]: @bennahugo The if statements above can probably be changed to continue in order to simply ignore FIELDS and SCANs

Comment (bennahugo) [reply xova/apps/xova/averaging.py]: @sjperkins I've changed this to continue in the commit. We could have a switch to flip between this. However, generally speaking unselecting scans and fields should remove them from the product.

As we discussed yesterday this is properly better handled through a taql query, but for the moment the current selection is the minimum needed for production use.

**Comment by sphemakh** [comment]
@sjperkins can we merge and release?

**Comment by sjperkins** [comment]
> @sjperkins can we merge and release?

No, this PR needs to change the selection to operate via TAQL

**Comment by sjperkins** [comment]
@SpheMakh Have you managed to test out the taql_where as per our email discussion?

<details><summary>email </summary>
<p>

MIME-Version: 1.0
Date: Tue, 28 Jan 2020 11:58:36 +0200
Message-ID: <CALazmLZeQLmh9rgfZdPPgGkuhjp4GkKddzNCm=zk8oUGw5LdzQ@mail.gmail.com>
Subject: taql_where
From: Simon Perkins <sperkins@ska.ac.za>
To: Sphesihle Makhathini <sphemakh@gmail.com>
Content-Type: multipart/alternative; boundary="000000000000878f32059d304449"

--000000000000878f32059d304449
Content-Type: text/plain; charset="UTF-8"

Look for the kwarg here

https://dask-ms.readthedocs.io/en/latest/api.html#daskms.xds_from_table

-- 
----------------------------------------------------------------
Simon Perkins
Radio Astronomy Research Group
South African Radio Astronomy Observatory
http://www.ska.ac.za
Tel: +27 (0)21 506 7369 <+27%2021%20506%207369>
----------------------------------------------------------------

--000000000000878f32059d304449
Content-Type: text/html; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

<div dir=3D"ltr"><div>Look for the kwarg here<br></div><div><br></div><div>=
<a href=3D"https://dask-ms.readthedocs.io/en/latest/api.html#daskms.xds_fro=
m_table">https://dask-ms.readthedocs.io/en/latest/api.html#daskms.xds_from_=
table</a></div><br>-- <br><div dir=3D"ltr" class=3D"gmail_signature" data-s=
martmail=3D"gmail_signature"><div dir=3D"ltr"><div><div dir=3D"ltr"><div><d=
iv dir=3D"ltr">------------------------------------------------------------=
----<br>Simon Perkins<br>Radio Astronomy Research Group<br>South African Ra=
dio Astronomy Observatory<br><a href=3D"http://www.ska.ac.za" target=3D"_bl=
ank">http://www.ska.ac.za</a><br>Tel: <a href=3D"tel:+27%2021%20506%207369"=
 value=3D"+27215067341" target=3D"_blank">+27 (0)21 506 7369</a><br>-------=
---------------------------------------------------------</div></div></div>=
</div></div></div></div>

--000000000000878f32059d304449--

</p>
</details>

**Comment by sphemakh** [comment]
> No, this PR needs to change the selection to operate via TAQL 

I see, I did not release. 

> @SpheMakh Have you managed to test out the taql_where as per our email discussion? 

Not, yet. Will get on this now

**Comment by sphemakh** [comment]
The CASA [MSselect syntax is supported by TaQL](http://www.aoc.nrao.edu/~sbhatnag/misc/msselection/UsingMSSexpressionsinTaQL.html#x23-3700010), so I can do something like this in CASA:

```
tb.taql("select from %s where mscal.field('0') || mscal.spw('0:1~10')" % vis)
```

But the equivalent raises a segmentation fault in dask-ms and casacore: 

```
dss = xds_from_ms(ms, taql_where="mscal.field('0') || mscal.spw('0:1~10')")
tab.query("mscal.field('0') || mscal.spw('0:1~10')") # casacore
```

The `mscal()` function is documented [here](https://casacore.github.io/casacore/classcasacore_1_1UDFMSCal.html)

**Comment by bennahugo** [comment]
Which version of python-casacore is this?

**Comment by bennahugo** [comment]
Could you compile python-cascore and casacore with debug symbols and get a stack trace with gdb and post the issue to casacore?

**Comment by sjperkins** [comment]
That functionality also segfaulted for me. I vaguely remember tracking down why but I can't remember now.

Does something like the following work?

```python
FIELD_ID IN [SELECT ROWID() FROM ::FIELD WHERE NAME IN ['3C286', 'PKS-1934']] AND
SCAN_NUMBER IN [0, 1, 2, 4] AND
DATA_DESC_ID == 0
```

**Comment by sphemakh** [comment]
That raises an exception:

```
RuntimeError: Error in TaQL command: 'using style Python SELECT
	FIELD_ID,
	DATA_DESC_ID,
	GAGGR(TIME) as GROUP_TIME,
	GROWID() AS __tablerow__,
	GCOUNT() as __tablerows__,
	GROWID()[0] as __firstrow__
FROM
	$1
GROUPBY
	FIELD_ID,
	DATA_DESC_ID
HAVING
	FIELD_ID IN [0] AND FIELD_ID IN [SELECT ROWID() FROM ::FIELD WHERE NAME IN ['3C286', 'PKS-1934']]'
  ColumnSet::doLock: table /home/sphe/work/github/xova/xova/workdir/1557766852_32chan.ms/FIELD should be locked when using UserLocking
```

But it goes away if open the `::FIELD` table before running the query 

```
ftab = table(ms+"::FIELD")
taql = "FIELD_ID IN [SELECT ROWID() FROM ::FIELD WHERE NAME IN ['3C286', 'PKS-1934']] AND SCAN_NUMBER IN [0, 1, 2, 4] AND DATA_DESC_ID == 0"
dss = xds_from_ms(ms, taql_where=taql)
ftab.close()
```

**Comment by bennahugo** [comment]
I think the more eloquent solution here is not to use subqueries and query the subtables first and inject the results into the query on the main table.

**Comment by sphemakh** [comment]
> I think the more eloquent solution here is not to use subqueries and query the subtables first and inject the results into the query on the main table. 

Agreed

**Comment by sphemakh** [comment]
In the meantime, for our data reductions, we can use CASA split to sub-spw selections of the data, and then use xova to average selected data.

**Comment by sjperkins** [comment]
> That raises an exception:
> 
> ```
> RuntimeError: Error in TaQL command: 'using style Python SELECT
> 	FIELD_ID,
> 	DATA_DESC_ID,
> 	GAGGR(TIME) as GROUP_TIME,
> 	GROWID() AS __tablerow__,
> 	GCOUNT() as __tablerows__,
> 	GROWID()[0] as __firstrow__
> FROM
> 	$1
> GROUPBY
> 	FIELD_ID,
> 	DATA_DESC_ID
> HAVING
> 	FIELD_ID IN [0] AND FIELD_ID IN [SELECT ROWID() FROM ::FIELD WHERE NAME IN ['3C286', 'PKS-1934']]'
>   ColumnSet::doLock: table /home/sphe/work/github/xova/xova/workdir/1557766852_32chan.ms/FIELD should be locked when using UserLocking
> ```
> 
> But it goes away if open the `::FIELD` table before running the query
> 
> ```
> ftab = table(ms+"::FIELD")
> taql = "FIELD_ID IN [SELECT ROWID() FROM ::FIELD WHERE NAME IN ['3C286', 'PKS-1934']] AND SCAN_NUMBER IN [0, 1, 2, 4] AND DATA_DESC_ID == 0"
> dss = xds_from_ms(ms, taql_where=taql)
> ftab.close()
> ```


Specifying the FIELD table independently. i.e specifying WSRT.MS/FIELD instead of WSRT.MS::FIELD or ::FIELD seems to work:

```python
ds = xds_from_ms("/home/sperkins/data/1527016443_sdp_l0.full_1284.full_pol.ms",
taql_where="FIELD_ID IN [SELECT ROWID() FROM /home/sperkins/data/1527016443_sdp_l0.full_1284.full_pol.ms/FIELD WHERE NAME IN ['3C286', 'PKS-1934']] AND SCAN_NUMBER IN [0, 1, 2, 4] AND DATA_DESC_ID == 0") 
```

I know this is verbose but it does provide access to the entire TAQL selection mechanism without reimplementing it.

**Pull request merged**